File: api_test_example.py
"""
闲鱼专属 API 调用示例 (Python 版本)

本文件提供了调用闲鱼专属 API 的完整示例代码
适用于 Python 环境下的外部软件集成

依赖:
    pip install requests
"""

import requests
import time
from typing import Dict, List, Optional

API_BASE_URL = 'http://localhost:3000'


class XianyuAPI:
    """闲鱼专属 API 客户端"""
    
    def __init__(self, base_url: str = API_BASE_URL):
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json'
        })
    
    def submit_account_email_info(self, account_data: Dict) -> Dict:
        """
        提交账号的 Outlook 邮箱信息到服务器
        
        Args:
            account_data: 账号数据字典，包含以下字段:
                - username: 账号用户名
                - password: 账号密码
                - outlook_email: Outlook 邮箱地址
                - mail_client_id: OAuth Client ID
                - mail_refresh_token: OAuth Refresh Token
        
        Returns:
            包含 success 和 account_id 或 error 的字典
        """
        try:
            response = self.session.post(
                f'{self.base_url}/api/xianyu/submit-email-info',
                json=account_data
            )
            response.raise_for_status()
            result = response.json()
            
            if result.get('success'):
                print(f'✅ 邮箱信息提交成功')
                print(f'账号ID: {result.get("account_id")}')
                return {
                    'success': True,
                    'account_id': result.get('account_id')
                }
            else:
                error_msg = result.get('message', '未知错误')
                print(f'❌ 提交失败: {error_msg}')
                return {
                    'success': False,
                    'error': error_msg
                }
                
        except requests.exceptions.RequestException as e:
            print(f'❌ 网络错误: {str(e)}')
            return {
                'success': False,
                'error': str(e)
            }
    
    def batch_submit_accounts(self, accounts: List[Dict], delay: float = 0.5) -> Dict:
        """
        批量提交多个账号的邮箱信息
        
        Args:
            accounts: 账号数据列表
            delay: 每次提交之间的延迟（秒）
        
        Returns:
            包含成功、失败统计的字典
        """
        print(f'开始批量提交 {len(accounts)} 个账号的邮箱信息...\n')
        
        results = {
            'success': 0,
            'failed': 0,
            'errors': []
        }
        
        for i, account in enumerate(accounts, 1):
            username = account.get('username', '未知')
            print(f'[{i}/{len(accounts)}] 提交账号: {username}')
            
            result = self.submit_account_email_info(account)
            
            if result['success']:
                results['success'] += 1
            else:
                results['failed'] += 1
                results['errors'].append({
                    'username': username,
                    'error': result.get('error', '未知错误')
                })
            
            # 避免请求过快
            if i < len(accounts):
                time.sleep(delay)
        
        print('\n============ 批量提交完成 ============')
        print(f'✅ 成功: {results["success"]}')
        print(f'❌ 失败: {results["failed"]}')
        
        if results['errors']:
            print('\n失败详情:')
            for err in results['errors']:
                print(f'  - {err["username"]}: {err["error"]}')
        
        return results
    
    def get_account_id(self, username: str) -> Optional[int]:
        """
        根据账号用户名获取账号ID
        
        Args:
            username: 账号用户名
        
        Returns:
            账号ID，如果失败则返回 None
        """
        try:
            response = self.session.post(
                f'{self.base_url}/api/xianyu/get-account-id',
                json={'username': username}
            )
            response.raise_for_status()
            result = response.json()
            
            if result.get('success'):
                return result.get('account_id')
            else:
                print(f'❌ 获取账号ID失败: {result.get("message")}')
                return None
                
        except requests.exceptions.RequestException as e:
            print(f'❌ 网络错误: {str(e)}')
            return None
    
    def get_email_verification_code(self, account_id: int) -> Optional[str]:
        """
        获取账号的邮箱验证码
        
        Args:
            account_id: 账号ID
        
        Returns:
            验证码，如果失败则返回 None
        """
        try:
            response = self.session.post(
                f'{self.base_url}/api/xianyu/get-email-code',
                json={'account_id': account_id}
            )
            response.raise_for_status()
            result = response.json()
            
            if result.get('success'):
                code = result.get('code')
                email = result.get('email')
                print(f'✅ 验证码获取成功: {code}')
                print(f'邮箱: {email}')
                return code
            else:
                error_msg = result.get('message', '未知错误')
                print(f'❌ 获取失败: {error_msg}')
                return None
                
        except requests.exceptions.RequestException as e:
            print(f'❌ 网络错误: {str(e)}')
            return None


def main():
    """主函数 - 演示 API 使用"""
    print('========== 闲鱼专属 API 测试 (Python) ==========\n')
    
    # 创建 API 客户端
    api = XianyuAPI()
    
    # 示例1: 提交单个账号
    print('示例1: 提交单个账号的邮箱信息')
    single_account = {
        'username': 'test@example.com',
        'password': 'password123',
        'outlook_email': 'test@outlook.com',
        'mail_client_id': 'your-client-id-here',
        'mail_refresh_token': 'your-refresh-token-here'
    }
    
    result = api.submit_account_email_info(single_account)
    print(f'提交结果: {result}\n')
    
    # 示例2: 批量提交多个账号
    print('\n示例2: 批量提交多个账号')
    accounts = [
        {
            'username': 'user1@example.com',
            'password': 'pass1',
            'outlook_email': 'user1@outlook.com',
            'mail_client_id': 'client-id-1',
            'mail_refresh_token': 'token-1'
        },
        {
            'username': 'user2@example.com',
            'password': 'pass2',
            'outlook_email': 'user2@outlook.com',
            'mail_client_id': 'client-id-2',
            'mail_refresh_token': 'token-2'
        }
    ]
    
    batch_results = api.batch_submit_accounts(accounts)
    
    # 示例3: 测试获取验证码（需要先提交过邮箱信息）
    print('\n\n示例3: 测试获取邮箱验证码')
    account_id = api.get_account_id('test@example.com')
    if account_id:
        code = api.get_email_verification_code(account_id)
        print(f'最终验证码: {code}')


def example_read_from_csv():
    """示例: 从 CSV 文件读取账号信息并批量提交"""
    import csv
    
    print('示例: 从 CSV 文件读取账号信息\n')
    
    # CSV 文件格式示例:
    # username,password,outlook_email,mail_client_id,mail_refresh_token
    # user1@example.com,pass1,user1@outlook.com,client-id-1,token-1
    # user2@example.com,pass2,user2@outlook.com,client-id-2,token-2
    
    csv_file = 'accounts.csv'
    
    try:
        accounts = []
        with open(csv_file, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                accounts.append({
                    'username': row['username'],
                    'password': row['password'],
                    'outlook_email': row['outlook_email'],
                    'mail_client_id': row['mail_client_id'],
                    'mail_refresh_token': row['mail_refresh_token']
                })
        
        print(f'从 {csv_file} 读取了 {len(accounts)} 个账号')
        
        api = XianyuAPI()
        results = api.batch_submit_accounts(accounts)
        
        print(f'\n✅ 成功提交 {results["success"]} 个账号')
        print(f'❌ 失败 {results["failed"]} 个账号')
        
    except FileNotFoundError:
        print(f'❌ 文件 {csv_file} 不存在')
    except Exception as e:
        print(f'❌ 读取文件时出错: {str(e)}')


def example_read_from_json():
    """示例: 从 JSON 文件读取账号信息并批量提交"""
    import json
    
    print('示例: 从 JSON 文件读取账号信息\n')
    
    # JSON 文件格式示例:
    # [
    #   {
    #     "username": "user1@example.com",
    #     "password": "pass1",
    #     "outlook_email": "user1@outlook.com",
    #     "mail_client_id": "client-id-1",
    #     "mail_refresh_token": "token-1"
    #   }
    # ]
    
    json_file = 'accounts.json'
    
    try:
        with open(json_file, 'r', encoding='utf-8') as f:
            accounts = json.load(f)
        
        print(f'从 {json_file} 读取了 {len(accounts)} 个账号')
        
        api = XianyuAPI()
        results = api.batch_submit_accounts(accounts)
        
        print(f'\n✅ 成功提交 {results["success"]} 个账号')
        print(f'❌ 失败 {results["failed"]} 个账号')
        
    except FileNotFoundError:
        print(f'❌ 文件 {json_file} 不存在')
    except Exception as e:
        print(f'❌ 读取文件时出错: {str(e)}')


if __name__ == '__main__':
    # 运行基本示例
    main()
    
    # 取消注释以下代码来测试从文件读取
    # example_read_from_csv()
    # example_read_from_json()

File: api_test_example.js
/**
 * 闲鱼专属 API 调用示例
 * 
 * 本文件提供了调用闲鱼专属 API 的完整示例代码
 * 适用于外部软件（如 Windows 桌面应用）集成
 */

const API_BASE_URL = 'http://localhost:3000';

// ============================================
// 1. 提交账号邮箱信息
// ============================================

/**
 * 提交账号的 Outlook 邮箱信息到服务器
 * @param {Object} accountData 账号数据
 * @returns {Promise<Object>} 提交结果
 */
async function submitAccountEmailInfo(accountData) {
    const {
        username,           // 账号用户名
        password,           // 账号密码
        outlook_email,      // Outlook 邮箱地址
        mail_client_id,     // OAuth Client ID
        mail_refresh_token  // OAuth Refresh Token
    } = accountData;

    try {
        const response = await fetch(`${API_BASE_URL}/api/xianyu/submit-email-info`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                username,
                password,
                outlook_email,
                mail_client_id,
                mail_refresh_token
            })
        });

        const result = await response.json();

        if (result.success) {
            console.log('✅ 邮箱信息提交成功');
            console.log('账号ID:', result.account_id);
            return { success: true, accountId: result.account_id };
        } else {
            console.error('❌ 提交失败:', result.message);
            return { success: false, error: result.message };
        }
    } catch (error) {
        console.error('❌ 网络错误:', error);
        return { success: false, error: error.message };
    }
}

// ============================================
// 2. 批量提交多个账号的邮箱信息
// ============================================

/**
 * 批量提交多个账号的邮箱信息
 * @param {Array<Object>} accounts 账号数组
 */
async function batchSubmitAccounts(accounts) {
    console.log(`开始批量提交 ${accounts.length} 个账号的邮箱信息...`);

    const results = {
        success: 0,
        failed: 0,
        errors: []
    };

    for (let i = 0; i < accounts.length; i++) {
        const account = accounts[i];
        console.log(`\n[${i + 1}/${accounts.length}] 提交账号: ${account.username}`);

        const result = await submitAccountEmailInfo(account);

        if (result.success) {
            results.success++;
        } else {
            results.failed++;
            results.errors.push({
                username: account.username,
                error: result.error
            });
        }

        // 避免请求过快，延迟 500ms
        await new Promise(resolve => setTimeout(resolve, 500));
    }

    console.log('\n============ 批量提交完成 ============');
    console.log(`✅ 成功: ${results.success}`);
    console.log(`❌ 失败: ${results.failed}`);

    if (results.errors.length > 0) {
        console.log('\n失败详情:');
        results.errors.forEach(err => {
            console.log(`  - ${err.username}: ${err.error}`);
        });
    }

    return results;
}

// ============================================
// 3. 测试获取邮箱验证码（可选）
// ============================================

/**
 * 根据账号用户名获取账号ID
 * @param {string} username 账号用户名
 * @returns {Promise<number|null>} 账号ID
 */
async function getAccountId(username) {
    try {
        const response = await fetch(`${API_BASE_URL}/api/xianyu/get-account-id`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ username })
        });

        const result = await response.json();
        return result.success ? result.account_id : null;
    } catch (error) {
        console.error('获取账号ID失败:', error);
        return null;
    }
}

/**
 * 获取账号的邮箱验证码
 * @param {number} accountId 账号ID
 * @returns {Promise<string|null>} 验证码
 */
async function getEmailVerificationCode(accountId) {
    try {
        const response = await fetch(`${API_BASE_URL}/api/xianyu/get-email-code`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ account_id: accountId })
        });

        const result = await response.json();

        if (result.success) {
            console.log('✅ 验证码获取成功:', result.code);
            console.log('邮箱:', result.email);
            return result.code;
        } else {
            console.error('❌ 获取失败:', result.message);
            return null;
        }
    } catch (error) {
        console.error('❌ 网络错误:', error);
        return null;
    }
}

// ============================================
// 使用示例
// ============================================

async function main() {
    console.log('========== 闲鱼专属 API 测试 ==========\n');

    // 示例1: 提交单个账号
    console.log('示例1: 提交单个账号的邮箱信息');
    const singleAccount = {
        username: 'test@example.com',
        password: 'password123',
        outlook_email: 'test@outlook.com',
        mail_client_id: 'your-client-id-here',
        mail_refresh_token: 'your-refresh-token-here'
    };

    const result = await submitAccountEmailInfo(singleAccount);
    console.log('提交结果:', result);

    // 示例2: 批量提交多个账号
    console.log('\n\n示例2: 批量提交多个账号');
    const accounts = [
        {
            username: 'user1@example.com',
            password: 'pass1',
            outlook_email: 'user1@outlook.com',
            mail_client_id: 'client-id-1',
            mail_refresh_token: 'token-1'
        },
        {
            username: 'user2@example.com',
            password: 'pass2',
            outlook_email: 'user2@outlook.com',
            mail_client_id: 'client-id-2',
            mail_refresh_token: 'token-2'
        }
    ];

    await batchSubmitAccounts(accounts);

    // 示例3: 测试获取验证码（需要先提交过邮箱信息）
    console.log('\n\n示例3: 测试获取邮箱验证码');
    const accountId = await getAccountId('test@example.com');
    if (accountId) {
        const code = await getEmailVerificationCode(accountId);
        console.log('最终验证码:', code);
    }
}

// 在 Node.js 环境中运行
if (typeof require !== 'undefined' && require.main === module) {
    // 需要安装 node-fetch: npm install node-fetch
    global.fetch = require('node-fetch');
    main().catch(console.error);
}

// 导出供其他模块使用
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        submitAccountEmailInfo,
        batchSubmitAccounts,
        getAccountId,
        getEmailVerificationCode
    };
}
